File: cli_go/core/commands.go
package core

import (
	"encoding/json"

	"github.com/charmbracelet/log"
	"github.com/robocorp/robo/cli/environment"
	"github.com/robocorp/robo/cli/process"
)

type Task struct {
	Name string
	Docs string
	File string
	Line int
}

func ListTasks(env environment.Environment) []Task {
	env.Variables["RC_LOG_OUTPUT_STDOUT"] = "1"

	proc := process.New("python", "-m", "robo", "list", "tasks.py")
	proc.Env = env.Variables

	output, err := proc.Run()
	if err != nil {
		log.Fatalf("Failed to read tasks: %v", output.Stderr)
	}

	var tasks []Task
	err = json.Unmarshal([]byte(output.Stdout), &tasks)
	if err != nil {
		log.Fatal(err)
	}

	return tasks
}

func RunTask(env environment.Environment, name string) {
	env.Variables["RC_LOG_OUTPUT_STDOUT"] = "1"

	proc := process.New("python", "-m", "robo", "run", "tasks.py", "-t", name)
	proc.Env = env.Variables
	proc.StdoutListener = func(line string) {
		log.Infof("Event: %v", line)
	}

	_, err := proc.Run()
	if err != nil {
		log.Fatalf("Failed to run task: %v", err)
	}
}

File: cli/src/robo_cli/core/commands.py
import json
from typing import Callable, TypedDict

from robo_cli.process import Process

from .events import TYPE_TO_EVENT, Event


class Task(TypedDict):
    name: str
    line: int
    file: str
    docs: str


def list_tasks(env: dict[str, str]) -> list[Task]:
    proc = Process(
        ["python", "-m", "robocorp.tasks", "list", "tasks.py"],
        env=env,
    )
    stdout, _ = proc.run()
    tasks = json.loads(stdout)
    return tasks


def run_task(env: dict[str, str], taskname: str, on_event: Callable[[Event], None]):
    proc = Process(
        ["python", "-m", "robocorp.tasks", "run", "tasks.py", "-t", taskname],
        env=env,
    )

    def handle_line(line: str):
        payload = json.loads(line)
        klass = TYPE_TO_EVENT[payload["message_type"]]
        event = klass.parse_obj(payload)
        on_event(event)

    proc.on_stdout(handle_line)
    proc.run()
